packages/core/src/routes/saml-assertion-handler.ts

```diff
@@ -0,0 +1,72 @@
+import type { ConnectorSession } from '@logto/connector-kit';
+import { ConnectorError, ConnectorErrorCodes, ConnectorType } from '@logto/connector-kit';
+import { arbitraryObjectGuard } from '@logto/schemas';
+import { z } from 'zod';
+
+import koaGuard from '#src/middleware/koa-guard.js';
+import assertThat from '#src/utils/assert-that.js';
+import {
+  getConnectorSessionResultFromJti,
+  assignConnectorSessionResultViaJti,
+} from '#src/utils/saml-assertion-handler.js';
+
+import type { AnonymousRouter, RouterInitArgs } from './types.js';
+
+export default function samlAssertionHandlerRoutes<T extends AnonymousRouter>(
+  ...[router, { provider, libraries }]: RouterInitArgs<T>
+) {
+  const {
+    socials: { getConnector },
+  } = libraries;
+
+  // Create an specialized API to handle SAML assertion
+  router.post(
+    '/saml-assertion-handler/:connectorId',
+    /**
+     * The API does not care the type of the SAML assertion request body, simply pass this to
+     * connector's built-in methods.
+     */
+    koaGuard({ body: arbitraryObjectGuard, params: z.object({ connectorId: z.string().min(1) }) }),
+    async (ctx, next) => {
+      const {
+        params: { connectorId },
+        body,
+      } = ctx.guard;
+      const connector = await getConnector(connectorId);
+      assertThat(connector.type === ConnectorType.Social, 'connector.unexpected_type');
+
+      const samlAssertionGuard = z.object({ SAMLResponse: z.string(), RelayState: z.string() });
+      const samlAssertionParseResult = samlAssertionGuard.safeParse(body);
+
+      if (!samlAssertionParseResult.success) {
+        throw new ConnectorError(
+          ConnectorErrorCodes.InvalidResponse,
+          samlAssertionParseResult.error
+        );
+      }
```

Comment:
assertThat?

Reply:
The linter can not properly infer that `.error` exists when parsing fails if `assert()` is applied.
<img width="1260" alt="image" src="https://user-images.githubusercontent.com/15182327/215307261-738c3602-63a1-4c91-a9f9-1db1e4ac2683.png">